rustls/src/crypto/aws_lc_rs/tls12.rs

```diff
@@ -0,0 +1,413 @@
+/// This implementation does the AAD construction required in TLS1.2.
+/// TLS1.3 uses `TLS13MessageDecrypter`.
+struct ChaCha20Poly1305MessageDecrypter {
+    dec_key: aead::LessSafeKey,
+    dec_offset: Iv,
+}
+
+const CHACHAPOLY1305_OVERHEAD: usize = 16;
+
+impl MessageDecrypter for ChaCha20Poly1305MessageDecrypter {
+    fn decrypt(&mut self, mut msg: OpaqueMessage, seq: u64) -> Result<PlainMessage, Error> {
+        let payload = msg.payload();
+
+        if payload.len() < CHACHAPOLY1305_OVERHEAD {
+            return Err(Error::DecryptError);
+        }
+
+        let nonce = aead::Nonce::assume_unique_for_key(Nonce::new(&self.dec_offset, seq).0);
+        let aad = aead::Aad::from(make_tls12_aad(
+            seq,
+            msg.typ,
+            msg.version,
+            payload.len() - CHACHAPOLY1305_OVERHEAD,
+        ));
+
+        let payload = msg.payload_mut();
+        let plain_len = self
+            .dec_key
+            .open_in_place(nonce, aad, payload)
+            .map_err(|_| Error::DecryptError)?
+            .len();
+
+        if plain_len > MAX_FRAGMENT_LEN {
+            return Err(Error::PeerSentOversizedRecord);
+        }
+
+        payload.truncate(plain_len);
+        Ok(msg.into_plain_message())
+    }
+}
+
+impl MessageEncrypter for ChaCha20Poly1305MessageEncrypter {
+    fn encrypt(&mut self, msg: BorrowedPlainMessage, seq: u64) -> Result<OpaqueMessage, Error> {
+        let nonce = aead::Nonce::assume_unique_for_key(Nonce::new(&self.enc_offset, seq).0);
+        let aad = aead::Aad::from(make_tls12_aad(seq, msg.typ, msg.version, msg.payload.len()));
+
+        let total_len = msg.payload.len() + self.enc_key.algorithm().tag_len();
+        let mut buf = Vec::with_capacity(total_len);
+        buf.extend_from_slice(msg.payload);
+
+        self.enc_key
+            .seal_in_place_append_tag(nonce, aad, &mut buf)
+            .map_err(|_| Error::EncryptError)?;
+
+        Ok(OpaqueMessage::new(msg.typ, msg.version, buf))
+    }
+}
+
+fn gcm_iv(write_iv: &[u8], explicit: &[u8]) -> Iv {
+    debug_assert_eq!(write_iv.len(), 4);
+    debug_assert_eq!(explicit.len(), 8);
+
+    // The GCM nonce is constructed from a 32-bit 'salt' derived
+    // from the master-secret, and a 64-bit explicit part,
+    // with no specified construction.  Thanks for that.
+    //
+    // We use the same construction as TLS1.3/ChaCha20Poly1305:
+    // a starting point extracted from the key block, xored with
+    // the sequence number.
+    let mut iv = [0; NONCE_LEN];
+    iv[..4].copy_from_slice(write_iv);
+    iv[4..].copy_from_slice(explicit);
+
+    Iv::new(iv)
+}
+
+struct Tls12Prf(&'static tls_prf::Algorithm);
+
+impl Prf for Tls12Prf {
+    fn for_secret(&self, output: &mut [u8], secret: &[u8], label: &[u8], seed: &[u8]) {
+        let result = tls_prf::Secret::new(self.0, secret)
+            .unwrap()
+            .derive(label, seed, output.len())
+            .unwrap();
```

Comment:
Well, hmm. First I think it's a mistake that we allow `export_keying_material()` to succeed with an empty output buffer; no good can come of that! I will add a commit which prevents that at a higher level and tests the behaviour.

I did do a cursory check that `CRYPTO_tls1_prf` (ie, `tls1_P_hash`) doesn't report failure on anything environment or otherwise in our control (eg, it doesn't use the allocator). It would be a bit of an imposition to make this fallible, I think, especially because it's such a small and well-defined deterministic function. I will comment the unwraps.